File: 02-two_pointers/00015-3Sum/00015-3sum.cpp
// 15: 3Sum
// https://leetcode.com/problems/3sum/

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    // SOLUTION
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> result;

        for (int i=0; i<nums.size(); i++) {
            if (i>0 && nums[i]==nums[i-1]) continue;

            int l = i+1;
            int r = nums.size() - 1;

            while (l<r) {
                int s = nums[i] + nums[l] + nums[r];
                if (s>0) r--;
                else if (s<0) l++;
                else {
                    result.push_back({nums[i], nums[l], nums[r]});
                    while (nums[l]==nums[l+1]) l++;
                    while (nums[r]==nums[r-1]) r--;
                    l++;
                    r--;
                }
            }
        }

        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> nums = {-1,0,1,2,-1,-4};

    // OUTPUT
    auto result = o.threeSum(nums);

    cout<<"["; for (auto x : result) {
        cout<<"["; for (auto y : x) cout<<y<<" "; cout<<"\b] ";
    } cout<<"\b]"<<endl;
    
    return 0;
}

File: 02-two_pointers/00015-3Sum/00015-3sum.py
# 15: 3Sum
# https://leetcode.com/problems/3sum/


class Solution:
    # SOLUTION
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        result: list[list[int]] = []

        for i in range(len(nums)):
            if i>0 and nums[i]==nums[i-1]: continue

            l = i+1
            r = len(nums) - 1

            while l<r:
                s = nums[i] + nums[l] + nums[r]
                if s>0: r-=1
                elif s<0: l+=1
                else:
                    result.append([nums[i], nums[l], nums[r]])
                    while nums[l]==nums[l+1]: l+=1
                    while nums[r]==nums[r-1]: r-=1
                    l+=1
                    r-=1
            
        return result



if __name__ == "__main__":
    o = Solution()

    # INPUT
    nums: list[int] = [-1,0,1,2,-1,-4]

    # OUTPUT
    result = o.threeSum(nums)
    print(result)